src/realm/util/future.hpp

```diff
@@ -525,6 +554,35 @@ class future_details::Promise {
     util::bind_ptr<SharedState<T>> m_shared_state = make_intrusive<SharedState<T>>();
 };
 
+/**
+ * CopyablePromiseHolder<T> is a lightweight copyable holder for Promises so they can be captured inside
+ * of std::function's and other types that require all members to be copyable.
+ *
+ * The only thing you can do with a CopyablePromiseHolder is extract a regular promise from it exactly once,
+ * and copy/move it as you would a util::bind_ptr.
+ *
+ * Do not use this type to try to fill a Promise from multiple places or threads.
+ */
+template <typename T>
+class future_details::CopyablePromiseHolder {
```

Comment:
We have a bunch of places in the codebase (mainly our tests) where we make a std::shared_ptr of a Promise so it can be captured in a std::function or other copyable type. This just lets us avoid that overkill by re-using the fact that the internal shared state of a Promise/Future is a ref-counted type.